### ISSUE TITLE
Issue in GPU concepts, NVIDIA architectures and branch divergence in kernel

### ISSUE THREAD

**Issue by hokkanen** [ENCCS/gpu-programming#27]
On 4-gpu-concepts.rst:124:

```
On some architectures, all members of a :abbr:`warp` have to execute the 
same instruction, so-called "lock-step" execution. This is done to achieve 
higher performance, but there are some drawbacks. If a an **if** statement 
is present inside a warp will cause the warp to be executed more than once, 
one time for each branch. On architectures without lock-step execution, such 
as NVIDIA Volta (e.g., GeForce 16xx-series) or newer, warp divergence is less costly.
```

To my understanding, GeForce 16xx-series is not an example of a NVIDIA Volta or newer. This might need be verified and potentially modified. I would also maybe clarify the claim about the if-statement; from what I've understood, there would be branch divergence only if the if-statement is evaluated at runtime (not templated branch), and multiple threads withing a single warp actually execute different branch of the if statement.

**Comment by al42and** [comment]
GeForce 16- and 20-series are Turing (sm_75), which is considered to be a consumer variant of Volta (sm_70), released ~2 years after the release of Volta. We can choose some other example, e.g. "TITAN V", the first consumer GPU with independent warp scheduling. I picked 16xx as they are (I suspect) more popular, and can be reasoned about: 16xx comes after 10xx but before 20xx (not quite accurate historically, but whatever), so 1080 has lock-step, 2070 does not.

> from what I've understood, there would be branch divergence only if the if-statement is evaluated at runtime (not templated branch), and multiple threads withing a single warp actually execute different branch of the if statement.

Agree.  How about the following:

> When different threads within a single :abbr:`warp` take different execution paths based on a conditional statement (**if**), both branches are executed sequentially, with some threads being active while others are inactive.

**Comment by stepas-toliautas** [comment]
@hokkanen , @al42and : any agreement on this? 😄 
My comments: a) if there is an arch example without lock-step, there should be a respective family/model name for the lock-step execution as well; and b) if every Turing is also Volta -- and if there is an incentive to use name "Volta" at all (this is due to Turing not being used much in "proper" HPC installations, right?) -- one can write "NVIDIA Volta **/ Turing** (e.g., GeForce 16xx-series)".

**Comment by al42and** [comment]
>  b) if every Turing is also Volta -- and if there is an incentive to use name "Volta" at all (this is due to Turing not being used much in "proper" HPC installations, right?) -- one can write "NVIDIA Volta / Turing (e.g., GeForce 16xx-series)".

Based on architecture numbers, Turing (sm_75) is "Volta (sm_70) **or newer**", but to avoid confusion, I guess "Volta / Turing" would be the best option.

**Comment by stepas-toliautas** [comment]
> Based on architecture numbers, Turing (sm_75) is "Volta (sm_70) **or newer**", but to avoid confusion, I guess "Volta / Turing" would be the best option.

So the point here is that lock-step free execution was introduced in Volta architecture and went on to stay (in NVidia's case)? I'm trying to understand whether the comparison represents some shift in thinking or if lock-step / no lock-step decisions are "to each their own".

**Comment by al42and** [comment]
> So the point here is that lock-step free execution was introduced in Volta architecture and went on to stay (in NVidia's case)?

Yes.

Implemented the following in 1f3f13f:

On some architectures, all members of a :abbr:`warp` have to execute the 
same instruction, the so-called "lock-step" execution. This is done to achieve 
higher performance, but there are some drawbacks. If an **if** statement 
is present inside a warp will cause the warp to be executed more than once, 
one time for each branch. When different threads within a single :abbr:warp
take different execution paths based on a conditional statement (if), both
branches are executed sequentially, with some threads being active while
others are inactive. On architectures without lock-step execution, such 
as NVIDIA Volta / Turing (e.g., GeForce 16xx-series) or newer, warp
divergence is less costly.
